src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -3,20 +3,19 @@ package org.kosat
 import kotlin.math.abs
 import kotlin.math.max
 
-// CDCL
-fun solveCnf(cnf: CnfRequest): List<Int>? {
-    val clauses = (cnf.clauses.map { it.lit }).toMutableList()
-    val solver = Kosat(clauses, cnf.vars)
-    return if (solver.solve()) solver.getModel() else null
-}
+abstract class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
+
+    abstract fun solve(): List<Int>?
+
+    protected abstract fun getNextVariable(level: Int): Int
 
-class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
     var varsNumber = initNumber
-        private set
+        protected set
 
     init {
```

Comment:
I would make `CDCL` class' primary constructor empty, and add another (secondary) ctor with `clauses: List<Clause>` argument. This secondary ctor would call the primary one (just `CDCL()`) and then `addClause` (+maybe `newVar`, if necessary) all the clauses passed in it.
That way your solver could be initialized empty (using primary ctor `CDCL()`) or "auto-filled" (using secondary ctor).

Moreover, consider replacing this "secondary ctor" with a static factory method (in Kotlin terms, method of the companion object). See "Item 30: Consider factory functions instead of constructors" in *Effective Kotlin* book.